File: challenges/birthday_bar.c
#include <stdio.h>
#include <stdlib.h>


/**
 *  count all the slices in the array
 *  where it's sum == sum_target
 *
 *  does not calc the sum for every slice
 *  instead
 *      calc the sum for the first slice
 *      do a rolling sum for all other slices
 *
 *  birthday(6, [2,3,2,7,1,4], 5, 2) == 3
 */
int birthday(const size_t s_size, const int* s, const int sum_target, const int length) {
    int sum = 0;
    int count = 0;

    for (size_t i = 0; i < length; ++i) {
        sum += s[i];
    }
    if (sum == sum_target) {
        ++count;
    }

    for (size_t i = 1; i < s_size - (length-1); ++i) {
        sum = sum - s[i-1] + s[i+(length-1)];
        if (sum == sum_target) {
            ++count;
        }
    }

    return count;
}


int main() {
    int squares = 0;
    scanf("%d", &squares);
    
    int * numbers = malloc(sizeof(int) * squares);
    
    for (int i=0; i<squares; ++i) {
        scanf("%d", &(numbers[i]));
    }

    int day = 0;
    scanf("%d", &day);
    int month = 0;
    scanf("%d", &month);

    const int result = birthday(squares, numbers, day, month);

    free(numbers);

    FILE* fptr = fopen(getenv("OUTPUT_PATH"), "w");
    fprintf(fptr, "%d\n", result);
    fclose(fptr);

    return 0;
}

File: challenges/BirthdayBar.hs
module Main where

import System.Environment.Blank
import System.IO


uncurry3 :: ([a] -> b -> c -> d) -> (([a], b, c) -> d)
uncurry3 f (x, y, z) = f x y z

fst3 :: (a, b, c) -> a
fst3 (x, _, _) = x


-- | count all the slices in the array
-- | where it's sum == sumTarget
--
--  does not calc the sum for every slice
--  instead
--      calc the sum for the first slice
--      do a rolling sum for all other slices
--
-- > birthday [5,1,4,6,2,3] 5 2 == 2
birthday :: [Int] -> Int -> Int -> Int
birthday numbers sumTarget sizeSlice = fstCount + counted
  where
  fstSlice = take sizeSlice numbers
  fstCount
    | sum fstSlice == sumTarget = 1
    | otherwise = 0

  counted = fst3 . foldl go (0, fstSlice, sum fstSlice) $ drop sizeSlice numbers
  go (count, slice, sumSlice) elem
    | sumR == sumTarget = (count+1, sliceR, sumR)
    | otherwise = (count, sliceR, sumR)
    where
    sliceR = tail slice ++ [elem]
    sumR = sumSlice - head slice + elem


output :: Int -> IO()
output result = do
  outFilePath <- getEnvDefault "OUTPUT_PATH" "/dev/stdout"
  withFile outFilePath WriteMode (\fptr -> do
          hPutStrLn fptr . show $ result
      )


input :: IO( ([Int], Int, Int) )
input = do
  _ <- readLn :: IO Int
  
  numbers <-
    getLine >>= return . map read . words
  
  (sliceSum, sliceLen) <-
    getLine >>= return . (\[x,y] -> (read x, read y)) . words

  return (numbers, sliceSum, sliceLen)


main :: IO()
main = do
  input >>= output . uncurry3 birthday
